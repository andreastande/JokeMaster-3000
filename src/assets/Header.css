.header {
  position: fixed;
  top: 0;
  z-index: 1;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 15vh;
  width: calc(100% - 8vw);
  padding: 0 4vw;
}

.dropdown-menu {
  display: flex;
  justify-content: center;
  width: 25%;
}

.logo {
  display: flex;
  justify-content: center;
  font-family: Orbitron-Bold;
  font-size: 5vh;
  width: 50%;
}

.toggle-wrapper {
  width: 25%;
  display: flex;
  justify-content: center;
}

.toggle {
  appearance: none;
  background-color: #f4f9fe;
  width: 50px;
  height: 50px;
  cursor: pointer;
  transition: 1s;
  border-radius: 100px;
  padding: 10px 13px;
  box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.25);
}

.dark .toggle {
  background-color: #4d6272;
}

.toggle:hover {
  background-color: rgb(43, 43, 43);
}

.toggle:after {
  content: "\1F319";
  font-size: 25px;
}

.toggle:checked:after {
  content: "\1F506";
}

@media (max-width: 1125px) {
  .header {
    height: 10vh;
  }

  .dropdownMenu {
    width: 30%;
  }

  .logo {
    flex-direction: column;
    align-items: center;
    width: 40%;
    font-size: 3.3vw;
  }

  .logo p {
    margin: 0;
  }

  .toggleWrapper {
    width: 30%;
    display: flex;
    justify-content: center;
  }
}

@media (max-width: 550px) {
  .toggle-wrapper {
    justify-content: flex-start;
    width: 20%;
  }

  .dropdown-menu {
    width: 40%;
  }
}

@media (max-width: 450px) {
  .logo {
    font-size: 12px;
  }
}

.dark .dropdown-button {
  background-color: #4d6272;
  border-color: #5a7387;
  color: #f4f9fe;
}

.dark .dropdown {
  background-color: #4d6272;
  border-color: #5a7387;
  color: #f4f9fe;
}
