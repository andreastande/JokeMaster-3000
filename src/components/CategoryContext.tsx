// CategoryContext.tsx
import React, { createContext, useState, useContext, ReactNode } from "react";

interface CategoryContextType {
  selectedCategory: Category;
  changeCategory: (category: Category) => void;
}

const Categories = {
  Programming: "Programming",
  Pun: "Pun",
  Spooky: "Spooky",
  Christmas: "Christmas",
  Favorites: "Favorites",
  Category: "Category",
} as const;

type Category = (typeof Categories)[keyof typeof Categories];

const CategoryContext = createContext<CategoryContextType | undefined>(
  undefined,
);

export const CategoryProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [selectedCategory, setSelectedCategory] = useState<Category>(
    Categories.Programming,
  );

  const changeCategory = (category: Category) => {
    setSelectedCategory(category);
  };

  return (
    <CategoryContext.Provider value={{ selectedCategory, changeCategory }}>
      {children}
    </CategoryContext.Provider>
  );
};

export const useCategory = (): CategoryContextType => {
  const context = useContext(CategoryContext);
  if (!context) {
    throw new Error("useCategory must be used within a CategoryProvider");
  }
  return context;
};
