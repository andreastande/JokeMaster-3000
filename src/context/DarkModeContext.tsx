import { createContext, useEffect, useState } from "react";

const DarkModeContext = ({ children }: { children: JSX.Element }) => {
  const [darkMode, setDarkMode] = useState(
    sessionStorage.getItem("darkMode") === "dark",
  );

  useEffect(() => {
    const html = document.documentElement;

    if (darkMode) {
      html.classList.add("dark");
    } else {
      html.classList.remove("dark");
    }
  }, [darkMode]);

  const newDarkMode = (updatedDarkMode: boolean) => {
    if (updatedDarkMode) {
      sessionStorage.setItem("darkMode", "dark");
    } else {
      sessionStorage.removeItem("darkMode");
    }
    setDarkMode(updatedDarkMode);
  };

  return (
    <DarkModeProvider.Provider value={{ darkMode, setDarkMode: newDarkMode }}>
      {children}
    </DarkModeProvider.Provider>
  );
};

export const DarkModeProvider = createContext({
  darkMode: false,
  setDarkMode: (() => {
    throw new Error("No DarkModeProvider");
  }) as (_: boolean) => void,
});

export default DarkModeContext;
