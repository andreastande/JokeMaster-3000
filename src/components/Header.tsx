import "../assets/Header.css";
import { DarkModeProvider } from "../context/DarkModeContext"; // Importing DarkModeContext
import { useContext } from "react";
import DropdownMenu from "./DropdownMenu";

type HeaderProps = {
  onSelect: (option: string) => void;
};

function Header({ onSelect }: HeaderProps) {
  const { darkMode, setDarkMode } = useContext(DarkModeProvider);
  const categories = ["Programming", "Pun", "Spooky", "Christmas", "Favorites"];

  const handleFilterChange = (value: string) => {
    onSelect(value);
  };

  return (
    <>
      <div className="header">
        <div className="dropdownMenu">
          <DropdownMenu
            filter="Category"
            options={categories}
            onSelect={(value) => handleFilterChange(value)}
          />
        </div>
        <div className="logo">
          <p>JOKEMASTER-</p>
          <p>3000</p>
        </div>
        <div className="toggleWrapper">
          <input
            className="toggle"
            type="checkbox"
            checked={darkMode}
            onClick={(e) => e.stopPropagation()}
            onChange={(e) => {
              setDarkMode(e.target.checked);
            }}
          />
        </div>
      </div>
    </>
  );
}

export default Header;
