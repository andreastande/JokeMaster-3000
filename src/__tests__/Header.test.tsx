import { act, render, screen } from '@testing-library/react';
import Header from '../components/Header';
import { test, expect} from 'vitest';
import { CategoryProvider } from '../context/CategoryContext';
import userEvent from '@testing-library/user-event';


test('Show dropdown when clicked', async () => {
    const {getByRole} = render(
      <CategoryProvider>
          <Header/>
      </CategoryProvider>
    );
    const dropDown = getByRole("button");
    await userEvent.click(dropDown);
    const programming = screen.getByText("Programming");
    expect(programming).toBeTruthy(); // The text is found
    const pun = screen.getByText("Pun");
    expect(pun).toBeTruthy(); // The text is found
    const favorites = screen.getByText("Favorites");
    expect(favorites).toBeTruthy(); // The text is found
})

test('Title is found', () => {
    const {getByText} = render(
      <CategoryProvider>
          <Header/>
      </CategoryProvider>
    );
    const title = getByText("JOKEMASTER-3000");
    expect(title).toBeTruthy(); // The text is found
})