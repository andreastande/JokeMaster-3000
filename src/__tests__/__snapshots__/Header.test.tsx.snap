// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Show dropdown when clicked 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <div
      class="dropdownButtonWrapper"
    >
      <button
        class="dropdownButton"
      >
        <label
          class="DdBlabel"
        >
          Category
        </label>
        <i
          class="dropdownArrow"
        />
      </button>
      <div
        class="dropdown"
      >
        <label>
          <input
            type="checkbox"
          />
          Programming
        </label>
        <label>
          <input
            type="checkbox"
          />
          Pun
        </label>
        <label>
          <input
            type="checkbox"
          />
          Spooky
        </label>
        <label>
          <input
            type="checkbox"
          />
          Christmas
        </label>
        <label>
          <input
            type="checkbox"
          />
          Favorites
        </label>
      </div>
    </div>
    <div
      class="logo"
    >
      <p>
        JOKEMASTER-3000
      </p>
    </div>
    <div
      class="toggleWrapper"
    >
      <input
        class="toggle"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;
